src/validate-icon.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright © 2018 Red Hat, Inc
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Matthias Clasen <mclasen@redhat.com>
+ */
+
+#include <gdk-pixbuf/gdk-pixbuf.h>
+#include <glib/gstdio.h>
+#include <errno.h>
+#include <unistd.h>
+
+static int
+validate_icon (const char *arg_width,
+               const char *arg_height,
+               const char *filename)
+{
+  GdkPixbufFormat *format;
+  int max_width, max_height;
+  int width, height;
+  const char *name;
+  const char *allowed_formats[] = { "png", "jpeg", "svg", NULL };
+  g_autoptr(GdkPixbuf) pixbuf = NULL;
+  g_autoptr(GError) error = NULL;
+
+  format = gdk_pixbuf_get_file_info (filename, &width, &height);
+  if (format == NULL)
+    {
+      g_printerr ("Format not recognized\n");
+      return 1;
+    }
+
+  name = gdk_pixbuf_format_get_name (format);
+  if (!g_strv_contains (allowed_formats, name))
+    {
+      g_printerr ("Format %s not accepted\n", name);
+      return 1;
+    }
+
+  if (!g_str_equal (name, "svg"))
+    {
+      max_width = g_ascii_strtoll (arg_width, NULL, 10);
+      if (max_width < 16 || max_width > 4096)
+        {
+          g_printerr ("Bad width limit: %s\n", arg_width);
+          return 1;
+        }
+
+      max_height = g_ascii_strtoll (arg_height, NULL, 10);
+      if (max_height < 16 || max_height > 4096)
+        {
+          g_printerr ("Bad height limit: %s\n", arg_height);
+          return 1;
+        }
+    }
+  else
+    {
+      /* Sanity check for vector files */
+      max_height = max_width = 4096;
+    }
+
+  if (width > max_width || height > max_height)
+    {
+      g_printerr ("Image too large (%dx%d). Max. size %dx%d\n", width, height, max_width, max_height);
+      return 1;
+    }
+
+  pixbuf = gdk_pixbuf_new_from_file (filename, &error);
```

Comment:
And this will read it a second time.

You might be better off reimplementing this loop using a loader and waiting to fetch the size yourself, depending on the possibly malicious image.